nibabel/streamlines/tck.py

```diff
@@ -0,0 +1,440 @@
+                self._write_header(f, header)
+
+                # Add the EOF_DELIMITER.
+                f.write(asbytes(self.EOF_DELIMITER.tostring()))
+                return
+
+            data_for_streamline = first_item.data_for_streamline
+            if len(data_for_streamline) > 0:
+                keys = ", ".join(data_for_streamline.keys())
+                msg = ("TCK format does not support saving additional data"
+                       " alongside streamlines. Dropping: {}".format(keys))
+                warnings.warn(msg, DataWarning)
+
+            data_for_points = first_item.data_for_points
+            if len(data_for_points) > 0:
+                keys = ", ".join(data_for_points.keys())
+                msg = ("TCK format does not support saving additional data"
+                       " alongside points. Dropping: {}".format(keys))
+                warnings.warn(msg, DataWarning)
+
+            # Make sure streamlines are in rasmm.
+            tractogram = self.tractogram.to_world(lazy=True)
+
+            for t in tractogram:
+                s = t.streamline.astype(dtype)
+                data = np.r_[s, self.FIBER_DELIMITER]
+                f.write(asbytes(data.tostring()))
+                nb_streamlines += 1
+
+            header[Field.NB_STREAMLINES] = nb_streamlines
+
+            # Add the EOF_DELIMITER.
+            f.write(asbytes(self.EOF_DELIMITER.tostring()))
+
+            # Overwrite header with updated one.
+            f.seek(beginning, os.SEEK_SET)
+            self._write_header(f, header)
+
+    @staticmethod
+    def _write_header(fileobj, header):
+        """ Write TCK header to file-like object.
+
+        Parameters
+        ----------
+        fileobj : file-like object
+            An open file-like object pointing to TCK file (and ready to read
+            from the beginning of the TCK header).
+        """
+        # Fields to exclude
+        exclude = [Field.MAGIC_NUMBER,  # Handled separately.
+                   Field.NB_STREAMLINES,  # Handled separately.
+                   Field.ENDIANNESS,  # Handled separately.
+                   Field.VOXEL_TO_RASMM,  # Streamlines are always in RAS+ mm.
+                   "count", "datatype", "file"]  # Fields being replaced.
+
+        lines = []
+        lines.append(header[Field.MAGIC_NUMBER])
+        lines.append("count: {0:010}".format(header[Field.NB_STREAMLINES]))
+        lines.append("datatype: Float32LE")  # Always Float32LE.
+        lines.extend(["{0}: {1}".format(k, v)
+                      for k, v in header.items() if k not in exclude and not k.startswith("_")])
+        lines.append("file: . ")  # Manually add this last field.
+        out = "\n".join((asstr(line).replace('\n', '\t') for line in lines))
+        fileobj.write(asbytes(out))
+
+        # Compute offset to the beginning of the binary data.
+        # We add 5 for "\nEND\n" that will be added just before the data.
+        tentative_offset = len(out) + 5
```

Comment:
Logic a bit tough to follow, and it's a bit hard to think about the assumptions that went into the `+1` case below.  How about:

```
hdr_len_no_offset = len(out) + 5
# Need to add number of bytes to store offset as decimal string. We start with estimate without string,
# then update if the offset-as-decimal-string got longer after adding length of the offset string. 
new_offset = -1
old_offset = hdr_len_no_offset
while new_offset != old_offset:
    old_offset = new_offset
    new_offset = hdr_len_no_offset + len(str(old_offset))
```